### DISCUSSION TITLE
Huggingface GGUF Editor

### DISCUSSION THREAD

**Discussion by CISC** [ggml-org/ggml#944] [Show and tell]
# [The Huggingface GGUF Editor](https://huggingface.co/spaces/CISCai/gguf-editor)
## 🎉 Check out my latest project 🌍✨

A powerful editor designed specifically for editing GGUF metadata and downloading the result directly from any Huggingface repository you have access to (you must sign in for access to gated or private ones).

With its user-friendly design, you can effortlessly edit any GGUF metadata through the [GGUF Editor](https://huggingface.co/spaces/CISCai/gguf-editor) hosted on Huggingface Spaces! 🌍✨🎉

Here are some basic usage examples:

* **Missing Pre-Tokenizer**: If you notice that there's missing/incorrect metadata like the crucial Pre-Tokenizer name, Fill-in-Middle tokens, or any other metadata based feature, edit/create GGUF metadata in a flash! 📝✨🔗

* **Chat Template Editing**: Suppose you want to fix an issue with the chat template(s) by updating them directly in the GGUF before downloading it, go ahead! Edit the existing template(s) or even add new ones for RAG, Tools, etc... 🗣️👥

* **Downloading Edited Files**: Once you've made your edits using the [GGUF Editor](https://huggingface.co/spaces/CISCai/gguf-editor), download your edited file directly from the Huggingface repository, no need to learn how to install and use complicated scripts! 🚀✨

![GGUF Editor](https://huggingface.co/spaces/CISCai/gguf-editor/resolve/main/ggufeditor.gif)

**Comment by julien-c** [comment]
very nice! Pinging @Vaibhavs10 as this might be of interest.

**Reply by CISC** [reply]
> very nice!

Thank you! I'm planning on making further improvements, at least partially supporting arrays to be able to change token types, maybe even fully, we'll see...

Another nice feature to have would be committing changes to your own repo and/or cloning whole repos with changes applied, however since that does not really seem practical with the limited storage capacity of the free plan it doesn't seem likely right now.

**Reply by Vaibhavs10** [reply]
Hey @CISC - Congrats on sucg a brilliant space - we'd be more than happy to support your mission, I started a discussion here: https://huggingface.co/spaces/CISCai/gguf-editor/discussions/1

**Reply by CISC** [reply]
Thank you very much! :)

**Comment by CISC** [comment]
I have just added support for easily editing individual tokenizer.ggml.scores/token_type entries with quick and easy lookup of every token, making this the most advanced GGUF Editor currently available. :)

This comes in real handy when models are released with incorrect metadata just like the recent Yi-Coder models that had incorrect token_type/score for the <|im_start|> token (as well as incorrect EOS token), there are still GGUFs with wrong metadata available...

**Comment by CISC** [comment]
Unfortunately there was an oversight that could cause unaligned tensors, this should now be fixed.

If you experienced nonsensical results from an edited GGUF, please grab a new copy, sorry for the inconvenience! :(
